tools/docker/lib/common

```diff
@@ -31,22 +28,94 @@ export CERTBOT_PLUGINS=(
     "dns-sakuracloud"
 )
 
-# Parses the requested architecture string and sets ALL_REQUESTED_ARCH to
-# result.
-# Usage: ParseRequestedArch [all|amd64|arm32v6|arm64v8]
-ParseRequestedArch() {
+# WORK_DIR is two levels above this file
+export WORK_DIR="$(realpath "$(dirname "${BASH_SOURCE[0]}")/..")"
+# REPO_ROOT is two levels above that
+export REPO_ROOT="$(realpath "${WORK_DIR}/../..")"
+
+# Converts input architecture identifier to the platform specification
+# understood by `docker build buildx --platform <specification>`.
+# Usage: arch2platform [arm64|arm32v6|arm64v8]
+# If the input is not recognized, an error is returned
+arch2platform() {
     REQUESTED_ARCH="${1}"
-    if [[ "${REQUESTED_ARCH}" == "all" ]]; then
-        ALL_REQUESTED_ARCH=("${ALL_TARGET_ARCH[@]}")
+    case $REQUESTED_ARCH in
+        amd64)
+            echo "linux/amd64"
+            ;;
+        arm32v6)
+            echo "linux/arm/v6"
+            ;;
+        arm64v8)
+            echo "linux/arm64"
+            ;;
+        *)
+            return 1
+            ;;
+    esac
+}
+
+
+# Interpret the user input. Expands "all" to a list of known arches,
+# and validates arches if provided with a list
+InterpretArchRequest() {
+    USER_INPUT="${1}"
+    local IFS=","
+    # Handle the special value "all"
+    if [[ "${USER_INPUT}" == "all" ]]; then
+        # Recursive call using the list of all known architectures cast to
+        # comma separated list
+        echo "${ALL_TARGET_ARCH[*]}"
         return 0
     fi
-    for TARGET_ARCH in "${ALL_TARGET_ARCH[@]}"; do
-        if [[ "${TARGET_ARCH}" == "${REQUESTED_ARCH}" ]]; then
-            ALL_REQUESTED_ARCH=("${REQUESTED_ARCH}")
-            return 0
+    # Convert comma separated list to array of strings
+    read -ra REQUESTED_ARCH_ARRAY <<< "$USER_INPUT"
+    for REQUESTED_ARCH in "${REQUESTED_ARCH_ARRAY[@]}"; do
+        local IFS=" "
+        if [[ ! " ${ALL_TARGET_ARCH[*]} " =~ " ${REQUESTED_ARCH} " ]]; then
+            echo "unknown architecture identifier: ${REQUESTED_ARCH}"
+            exit 1
         fi
     done
-    # If we didn't return above, REQUESTED_ARCH has an unexpected value.
-    echo "Unexpected target architecture \"${REQUESTED_ARCH}\"". >&2
-    exit 1
+    echo "$USER_INPUT"
+}
+
+ParseArgs() {
+
+    export TAG_VER="$1"
+    if [ -z "$TAG_VER" ]; then
+        echo "We cannot tag Docker images with an empty string!" >&2
+        exit 1
+    fi
+    ARCH_LIST="$2"
+    if [ -z "$ARCH_LIST" ]; then
+        echo "Architectures must be specified!" >&2
+        exit 1
+    fi
+    # split arch list into an array for per-arch image building and saving
+    local IFS=","
+    read -ra REQUESTED_ARCH_ARRAY <<< $(InterpretArchRequest "$ARCH_LIST")
```

Comment:
I think there's a few ways to fix this, but I noticed this masks any failures in `InterpretArchRequest`. It exiting is ignored and the script continues to execute. Can we fix that?

Reply:
Pulled the function out of a subshell and put the logic directly in the caller. It was the only invocation so it actually made things more compact.

Reply:
fixed in https://github.com/certbot/certbot/pull/9628/commits/537ef57b0a155144850ba07676ae90802311bc80